pkg/ingester/circuitbreaker.go

```diff
@@ -0,0 +1,227 @@
+		}).
+		OnOpen(func(event circuitbreaker.StateChangedEvent) {
+			transitionOpen.Inc()
+			level.Info(ingester.logger).Log("msg", "circuit breaker is open", "ingester", ingesterID, "previous", event.OldState, "current", event.NewState)
+		}).
+		OnHalfOpen(func(event circuitbreaker.StateChangedEvent) {
+			transitionHalfOpen.Inc()
+			level.Info(ingester.logger).Log("msg", "circuit breaker is half-open", "ingester", ingesterID, "previous", event.OldState, "current", event.NewState)
+		}).
+		HandleIf(func(_ any, err error) bool { return isFailure(err) })
+
+	if cfg.testModeEnabled {
+		// In case of testing purposes, we initialize the circuit breaker with count based failure thresholding,
+		// since it is more deterministic, and therefore it is easier to predict the outcome.
+		cbBuilder = cbBuilder.WithFailureThreshold(cfg.FailureThreshold)
+	} else {
+		// In case of production code, we prefer time based failure thresholding.
+		cbBuilder = cbBuilder.WithFailureRateThreshold(cfg.FailureThreshold, cfg.FailureExecutionThreshold, cfg.ThresholdingPeriod)
+	}
+
+	cb := cbBuilder.Build()
+	return &circuitBreaker{
+		CircuitBreaker: cb,
+		ingester:       ingester,
+		executor:       failsafe.NewExecutor[any](cb),
+		startTime:      time.Now().Add(cfg.InitialDelay),
+	}
+}
+
+func isFailure(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	// We only consider timeouts or ingester hitting a per-instance limit
+	// to be errors worthy of tripping the circuit breaker since these
+	// are specific to a particular ingester, not a user or request.
+
+	if errors.Is(err, context.DeadlineExceeded) {
+		return true
+	}
+
+	var ingesterErr ingesterError
+	if errors.As(err, &ingesterErr) {
+		return ingesterErr.errorCause() == mimirpb.INSTANCE_LIMIT
+	}
+
+	return false
+}
+
+func (cb *circuitBreaker) isActive() bool {
+	return cb.startTime.Before(time.Now())
+}
+
+func (cb *circuitBreaker) ingesterID() string {
+	return cb.ingester.cfg.IngesterRing.InstanceID
+}
+
+func (cb *circuitBreaker) logger() log.Logger {
+	return cb.ingester.logger
+}
+
+func (cb *circuitBreaker) metrics() *ingesterMetrics {
+	return cb.ingester.metrics
+}
+
+func (cb *circuitBreaker) config() CircuitBreakerConfig {
+	return cb.ingester.cfg.CircuitBreakerConfig
+}
+
+func (cb *circuitBreaker) get(ctx context.Context, callbackFnName string, callbackFn func() (any, error)) (any, error) {
+	res, err := cb.executor.Get(callbackFn)
+	if err != nil && errors.Is(err, circuitbreaker.ErrOpen) {
+		cb.metrics().circuitBreakerResults.WithLabelValues(cb.ingesterID(), resultOpen).Inc()
+		cbOpenErr := middleware.DoNotLogError{Err: newCircuitBreakerOpenError(cb.RemainingDelay())}
+		return res, newErrorWithStatus(cbOpenErr, codes.Unavailable)
+	}
+	return res, cb.processError(ctx, err, callbackFnName)
+}
+
+func (cb *circuitBreaker) processError(ctx context.Context, err error, callbackFnName string) error {
```

Comment:
This method is too shallow, and can be merged into `get`.

Reply:
This PR is just one part of the work beacuse, in order to simplify the review, I decided to split the write from the read path. Since there will be different additional methods on the read path, and all of them will need to process these errors in the same way, I decided to add this method.